src/Uno.UI/UI/Xaml/Internal/EventManager.cs

```diff
@@ -0,0 +1,159 @@
+﻿#nullable enable
+
+#if UNO_HAS_ENHANCED_LIFECYCLE
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.UI.Xaml;
+using Microsoft.UI.Xaml.Controls;
+using Uno.UI.Dispatching;
+
+namespace Uno.UI.Xaml.Core;
+
+internal sealed class EventManager
+{
+	private List<DependencyObject?>? _loadedEventList;
+	private List<(FrameworkElement Element, EffectiveViewportChangedEventArgs Args)>? _effectiveViewportChangedQueue;
+
+	internal bool ShouldRaiseLoadedEvent { get; private set; }
+
+	internal bool HasPendingViewportChangedEvents => _effectiveViewportChangedQueue is { Count: > 0 };
+
+	private EventManager()
+	{
+	}
+
+	internal void EnqueueForEffectiveViewportChanged(FrameworkElement element, EffectiveViewportChangedEventArgs args)
+	{
+		if (_effectiveViewportChangedQueue is null)
+		{
+			_effectiveViewportChangedQueue = new();
+		}
+
+		_effectiveViewportChangedQueue.Add((element, args));
+		CoreServices.RequestAdditionalFrame();
+	}
+
+	internal void RaiseEffectiveViewportChangedEvents()
+	{
+		if (_effectiveViewportChangedQueue is null)
+		{
+			return;
+		}
+
+		for (int i = 0; i < _effectiveViewportChangedQueue.Count; i++)
+		{
+			var (fe, args) = _effectiveViewportChangedQueue[i];
+
+			_effectiveViewportChangedQueue[i] = default;
+
+			fe.RaiseEffectiveViewportChanged(args);
+		}
+
+		_effectiveViewportChangedQueue.Clear();
+
+	}
+
+	private void AddToLoadedEventList(DependencyObject element)
+	{
+		if (_loadedEventList is null)
+		{
+			_loadedEventList = new();
+		}
+
+		// TODO: Maybe this shouldn't happen?
+		// It doesn't make sense to attempt to add duplicate elements.
```

Comment:
Should we use a `HashSet` then? Does the ordering matters?

Reply:
I think the ordering may matter, yes.